### PR TITLE
download progress. take 2

### PR DESCRIPTION
**Pull request by michaelkirk**
PTAL @charlesmchen

### CODE REVIEW

File: Signal/src/Models/TSMessageAdapaters/AttachmentPointerAdapter.swift [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        JSQMessagesMediaViewBubbleImageMasker.applyBubbleImageMask(toMediaView: view, isOutgoing:!isIncoming)
+
+        view.isUserInteractionEnabled = false
+
+        let attachmentPointerView = AttachmentPointerView(attachmentPointer: attachmentPointer, isIncoming: self.isIncoming)
+        self.attachmentPointerView = attachmentPointerView
+        view.addSubview(attachmentPointerView)
+
+        attachmentPointerView.autoPinWidthToSuperview(withMargin: 20)
+        attachmentPointerView.autoVCenterInSuperview()
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/AttachmentPointerAdapter.swift]: So `AttachmentPointerView` is guaranteed to always be smaller than `view`? Should we have `view` clip to bounds?

File: Signal/src/Models/TSMessageAdapaters/TSMessageAdapter.m [hunk 2]
```diff
@@ -177,19 +177,9 @@ - (instancetype)initWithInteraction:(TSInteraction *)interaction
                     }
                 } else if ([attachment isKindOfClass:[TSAttachmentPointer class]]) {
                     TSAttachmentPointer *pointer = (TSAttachmentPointer *)attachment;
-                    adapter.messageType          = TSInfoMessageAdapter;
-
-                    switch (pointer.state) {
-                        case TSAttachmentPointerStateEnqueued:
-                            adapter.messageBody = NSLocalizedString(@"ATTACHMENT_QUEUED", nil);
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSMessageAdapter.m]: Did you happen to find out if we're queuing attachment downloads and if so what's the max number of simultaneous downloads?

File: Signal/src/ViewControllers/MessagesViewController.m [hunk 3]
```diff
@@ -1201,6 +1201,16 @@ - (BOOL)collectionView:(JSQMessagesCollectionView *)collectionView shouldShowMen
     // JSQM does some setup in super method
     [super collectionView:collectionView shouldShowMenuForItemAtIndexPath:indexPath];
 
+    id<OWSMessageData> message = [self messageAtIndexPath:indexPath];
+    if ([message.media isKindOfClass:[AttachmentPointerAdapter class]]) {
+        AttachmentPointerAdapter *attachmentPointerAdapter = (AttachmentPointerAdapter *)message.media;
+        if (attachmentPointerAdapter.attachmentPointer.state == TSAttachmentPointerStateFailed) {
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessagesViewController.m]: `return attachmentPointerAdapter.attachmentPointer.state == TSAttachmentPointerStateFailed;`

File: Signal/src/ViewControllers/MessagesViewController.m [hunk 4]
```diff
@@ -1856,15 +1866,38 @@ - (void)collectionView:(JSQMessagesCollectionView *)collectionView
                             }
                         }
                     }
+                } else if ([messageItem.media isKindOfClass:[AttachmentPointerAdapter class]]) {
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessagesViewController.m]: I really would like to decompose this method into a bunch of smaller methods.

File: Signal/src/ViewControllers/MessagesViewController.m [hunk 5]
```diff
@@ -1856,15 +1866,38 @@ - (void)collectionView:(JSQMessagesCollectionView *)collectionView
...
+                    AttachmentPointerAdapter *attachmentPointerAdadpter = (AttachmentPointerAdapter *)messageItem.media;
+
+                    TSAttachmentPointer *pointer = attachmentPointerAdadpter.attachmentPointer;
+                    if (pointer.state == TSAttachmentPointerStateFailed) {
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessagesViewController.m]: Does this "tap on download"?  Should it present an action sheet with "try again/cancel"? Just asking.

File: Signal/src/views/AttachmentPointerView.swift [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        super.init(frame: frame)
+        self.createSubviews()
+        self.updateViews()
+    }
+
+    @available(*, unavailable)
+    required init?(coder aDecoder: NSCoder) {
+        assertionFailure()
+        // These methods should never be called, but we
+        self.filename = "Attachment"
```
Comment (charlesmchen) [inline Signal/src/views/AttachmentPointerView.swift]: localize?

Comment (charlesmchen) [inline Signal/src/views/AttachmentPointerView.swift]: oh this shouldn't ever be called.

File: Signal/src/views/AttachmentPointerView.swift [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        self.attachmentPointer = TSAttachmentPointer()
+        super.init(frame: frame)
+        self.createSubviews()
+        self.updateViews()
+    }
+
+    @available(*, unavailable)
+    required init?(coder aDecoder: NSCoder) {
+        assertionFailure()
+        // These methods should never be called, but we
```
Comment (charlesmchen) [inline Signal/src/views/AttachmentPointerView.swift]: Incomplete comment.

File: Signal/src/Models/TSMessageAdapaters/AttachmentPointerAdapter.swift (hunk 1 above)
Comment (michaelkirk) [inline Signal/src/Models/TSMessageAdapaters/AttachmentPointerAdapter.swift]: done.

File: Signal/src/Models/TSMessageAdapaters/TSMessageAdapter.m (hunk 2 above)
Comment (michaelkirk) [inline Signal/src/Models/TSMessageAdapaters/TSMessageAdapter.m]: They're all *started* from a serial queue. But then they quickly get dispatched to a concurrent network queue. So no limit AFAIK.

File: Signal/src/ViewControllers/MessagesViewController.m (hunk 3 above)
Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: done.

File: Signal/src/ViewControllers/MessagesViewController.m (hunk 4 above)
Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: Probably not now, but smells ripe for some good ole [replace conditional with polymorphism](https://refactoring.com/catalog/replaceConditionalWithPolymorphism.html)

File: Signal/src/ViewControllers/MessagesViewController.m (hunk 5 above)
Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: It probably should, to be consistent with our "failed download" UI.

File: Signal/src/views/AttachmentPointerView.swift (hunk 6 above)
Comment (michaelkirk) [inline Signal/src/views/AttachmentPointerView.swift]: done

File: Signal/src/views/AttachmentPointerView.swift (hunk 7 above)
Comment (michaelkirk) [inline Signal/src/views/AttachmentPointerView.swift]: done

File: Signal/src/ViewControllers/MessagesViewController.m (hunk 5 above)
Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: done.

**Pull request merged**